File: day15/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
)

const (
	puzzleInput = "input.txt"
	puzzleRow   = 2000000
	puzzleBound = 4000000
)

func main() {
	lineRegex := regexp.MustCompile(`x=(-?\d+).*y=(-?\d+).*x=(-?\d+).*y=(-?\d+)`)
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var sensors []Sensor
	beacons := map[Pos]struct{}{}
	first := true
	var leftBound, rightBound int

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		matches := lineRegex.FindStringSubmatch(scanner.Text())
		if len(matches) == 0 {
			log.Fatalln("Invalid line")
		}
		x1, err := strconv.Atoi(matches[1])
		if err != nil {
			log.Fatalln(err)
		}
		y1, err := strconv.Atoi(matches[2])
		if err != nil {
			log.Fatalln(err)
		}
		x2, err := strconv.Atoi(matches[3])
		if err != nil {
			log.Fatalln(err)
		}
		y2, err := strconv.Atoi(matches[4])
		if err != nil {
			log.Fatalln(err)
		}
		pos := Pos{
			y: y1,
			x: x1,
		}
		beacon := Pos{
			y: y2,
			x: x2,
		}
		beacons[beacon] = struct{}{}
		radius := manhattanDistance(pos, beacon)
		sensor := Sensor{
			pos:    pos,
			radius: radius,
		}
		sensors = append(sensors, sensor)
		if x1, x2, ok := sensor.BoundsX(puzzleRow); ok {
			if first {
				first = false
				leftBound = x1
				rightBound = x2
			} else {
				if x1 < leftBound {
					leftBound = x1
				}
				if x2 > rightBound {
					rightBound = x2
				}
			}
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	{
		count := 0
		if !first {
			for x := leftBound; x <= rightBound; x++ {
				pos := Pos{
					y: puzzleRow,
					x: x,
				}
				if _, ok := beacons[pos]; ok {
					continue
				}
				for _, i := range sensors {
					if !i.InRange(pos) {
						continue
					}
					_, x2, ok := i.BoundsX(puzzleRow)
					if !ok {
						log.Fatalln("Invariant violated")
					}
					if x2 == x {
						count++
					} else {
						count += x2 - x
						x = x2 - 1
					}
					break
				}
			}
		}
		fmt.Println("Part 1:", count)
	}
	{
	outer2:
		for y := 0; y <= puzzleBound; y++ {
		outer:
			for x := 0; x <= puzzleBound; x++ {
				pos := Pos{
					y: y,
					x: x,
				}
				for _, i := range sensors {
					if !i.InRange(pos) {
						continue
					}
					_, x2, ok := i.BoundsX(y)
					if !ok {
						log.Fatalln("Invariant violated")
					}
					x = x2
					continue outer
				}
				fmt.Println("Part 2:", x*puzzleBound+y)
				break outer2
			}
		}
	}
}

type (
	Pos struct {
		y, x int
	}

	Sensor struct {
		pos    Pos
		radius int
	}
)

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func manhattanDistance(a, b Pos) int {
	return abs(a.x-b.x) + abs(a.y-b.y)
}

func (s *Sensor) InRange(pos Pos) bool {
	return manhattanDistance(s.pos, pos) <= s.radius
}

func (s *Sensor) BoundsX(y int) (int, int, bool) {
	vdelta := abs(s.pos.y - y)
	if vdelta > s.radius {
		return 0, 0, false
	}
	delta := s.radius - vdelta
	return s.pos.x - delta, s.pos.x + delta, true
}

File: day15/src/main.rs
use regex::Regex;
use std::collections::HashSet;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";
const PUZZLE_ROW: i32 = 2000000;
const PUZZLE_BOUND: i32 = 4000000;

type BoxError = Box<dyn std::error::Error>;
type BoxResult<T> = Result<T, BoxError>;

fn main() -> BoxResult<()> {
    let line_regex = Regex::new(r"x=(-?\d+).*y=(-?\d+).*x=(-?\d+).*y=(-?\d+)").unwrap();

    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut sensors = Vec::new();
    let mut beacons = HashSet::new();
    let mut bounds = None;

    for line in reader.lines() {
        let line = line?;
        let captures = line_regex.captures(&line).ok_or("Invalid line")?;
        let pos = Pos::new(
            captures.get(1).ok_or("Invalid line")?.as_str().parse()?,
            captures.get(2).ok_or("Invalid line")?.as_str().parse()?,
        );
        let beacon = Pos::new(
            captures.get(3).ok_or("Invalid line")?.as_str().parse()?,
            captures.get(4).ok_or("Invalid line")?.as_str().parse()?,
        );
        beacons.insert(beacon);
        let radius = pos.manhattan_distance(&beacon);
        let sensor = Sensor::new(pos, radius);
        if let Some((x1, x2)) = sensor.bounds_x(PUZZLE_ROW) {
            match &mut bounds {
                None => bounds = Some((x1, x2)),
                Some((v1, v2)) => {
                    if x1 < *v1 {
                        *v1 = x1
                    }
                    if x2 > *v2 {
                        *v2 = x2
                    }
                }
            }
        }
        sensors.push(sensor);
    }

    {
        let mut count = 0;
        if let Some((left_bound, right_bound)) = bounds {
            let mut x = left_bound;
            while x <= right_bound {
                let pos = Pos::new(x, PUZZLE_ROW);
                if beacons.contains(&pos) {
                    x += 1;
                    continue;
                }
                for i in &sensors {
                    if !i.in_range(&pos) {
                        continue;
                    }
                    let (_, x2) = i.bounds_x(PUZZLE_ROW).ok_or("Invariant violated")?;
                    if x2 == x {
                        count += 1;
                    } else {
                        count += x2 - x;
                        x = x2 - 1;
                    }
                    break;
                }
                x += 1;
            }
        }
        println!("Part 1: {}", count);
    }
    {
        let mut y = 0;
        'outer2: while y <= PUZZLE_BOUND {
            let mut x = 0;
            'outer: while x <= PUZZLE_BOUND {
                let pos = Pos::new(x, y);
                for i in &sensors {
                    if !i.in_range(&pos) {
                        continue;
                    }
                    let (_, x2) = i.bounds_x(y).ok_or("Invariant violated")?;
                    x = x2 + 1;
                    continue 'outer;
                }
                println!("Part 2: {}", x as i64 * PUZZLE_BOUND as i64 + y as i64);
                break 'outer2;
            }
            y += 1
        }
    }

    Ok(())
}

#[derive(Clone, Copy, PartialEq, Eq, Hash)]
struct Pos {
    y: i32,
    x: i32,
}

impl Pos {
    fn new(x: i32, y: i32) -> Self {
        Self { y, x }
    }

    fn manhattan_distance(&self, other: &Self) -> i32 {
        (self.x - other.x).abs() + (self.y - other.y).abs()
    }
}

struct Sensor {
    pos: Pos,
    radius: i32,
}

impl Sensor {
    fn new(pos: Pos, radius: i32) -> Self {
        Self { pos, radius }
    }

    fn in_range(&self, pos: &Pos) -> bool {
        self.pos.manhattan_distance(&pos) <= self.radius
    }

    fn bounds_x(&self, y: i32) -> Option<(i32, i32)> {
        let vdelta = (self.pos.y - y).abs();
        if vdelta > self.radius {
            None
        } else {
            let delta = self.radius - vdelta;
            Some((self.pos.x - delta, self.pos.x + delta))
        }
    }
}
